### PR TITLE
Runme E2E Tests improvements

### PR DESCRIPTION
**Pull request by degrammer**
Improve E2E tests. Still needs to fix some failing tests and organize better the code

### CODE REVIEW

File: tests/e2e/pageobjects/cell.page.ts [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+
+        for await (const row of successStateRows) {
+            const parent = await row
+                .parentElement()
+                .parentElement()
+                .parentElement()
+                .parentElement()
+                .parentElement()
+            const cellEditorContainer = await parent.$('.cell-editor-container')
+            if (await cellEditorContainer.getText() === this.#cellText) {
```
Comment (christian-bromann) [inline tests/e2e/pageobjects/cell.page.ts:47]: It seems like this looks very much like what is done in `isSuccessfulExecution`. Maybe we can have a function like:
```ts
getCurrentExecutionRow () {
    const successRows = await browser.$$('.codicon-notebook-state-success')
    for await (const row of successRows) {
        const executionRow = await row
            .parentElement()
            .parentElement()
            .parentElement()
            .parentElement()
            .parentElement()
        const cellEditor = await executionRow.$('.cell-editor-container')
        const text = await cellEditor.getText()
        if (text === this.#cellText) {
            return executionRow
        }
    }
}
```

and shrink this method down to:

```ts
const commands: NotebookCommand[] = []
const execRow = this.#getCurrentExecutionRow()
if (execRow) {
    return commands
}

const statusBar$ = await execRow.$('.cell-statusbar-container')
const statusRight$ = await statusBar$.$('.cell-status-right')
const contributedRight$ = await statusRight$.$('.cell-contributed-items-right')
const commandsResult$$ = await contributedRight$.$$('.cell-status-item-has-command')

for await (const row of commandsResult$$) {
    const text = await row.getText()
    commands.push({
        element$: row,
        text: text.trim()
    })
}

return commands
```

File: tests/e2e/pageobjects/cell.page.ts [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+                .parentElement()
+            const cellEditor = await executionRow.$('.cell-editor-container')
+            const text = await cellEditor.getText()
+            if (text === this.#cellText) {
+                cellExists = true
+                break
+            }
+        }
+        return cellExists
+    }
```
Comment (christian-bromann) [inline tests/e2e/pageobjects/cell.page.ts:98]: and this one to
```suggestion
async isSuccessfulExecution(): Promise<boolean> {
    return Boolean(await this.#getCurrentExecutionRow())
}
```

File: tests/e2e/specs/basic.e2e.ts [hunk 3]
```diff
@@ -1,6 +1,13 @@
 import { DefaultTreeItem } from 'wdio-vscode-service'
+import { Key } from 'webdriverio'
+import clipboard from 'clipboardy'
 
-describe('Runme VS Code Extension', () => {
+import { Notebook } from '../pageobjects/notebook.page.js'
+import { OutputType } from '../pageobjects/cell.page.js'
+
+
+
```
Comment (christian-bromann) [inline tests/e2e/specs/basic.e2e.ts:9]:
```suggestion
```

File: tests/e2e/specs/basic.e2e.ts [hunk 4]
```diff
@@ -51,27 +58,208 @@ describe('Runme VS Code Extension', () => {
...
+      await cell.run()
+      expect(await cell.cellOutputExists('', OutputType.ShellOutput)).toBe(true)
+      await cell.focus()
+    })
+
+    // TODO: Review why this is failing
+    it.skip('complex output', async () => {
+      const cell = await notebook.getCell('npm i -g webdriverio')
+      await cell.run()
+      const regex = new RegExp(/(added|changed) \d+ packages in \d+(?:ms|s)/)
```
Comment (christian-bromann) [inline tests/e2e/specs/basic.e2e.ts:95]: let's move this to the top of the file

**Review by christian-bromann — COMMENTED** [review]
Good stuff 👌 some comments

**Comment by degrammer** [comment]
@christian-bromann  fixed issues and organized the code better according to your suggestions.

File: tests/e2e/pageobjects/cell.page.ts [hunk 5]
```diff
@@ -0,0 +1,197 @@
...
+
+    async run() {
+        const container = await this.#cellRow.parentElement().parentElement()
+        await container.$('.run-button-container').click()
+    }
+
+    async #getStatusBarElements(): Promise<NotebookCommand[]> {
+        const executionRow = await this.#getCurrentExecutionRow()
+        let commands: NotebookCommand[] = []
```
Comment (christian-bromann) [inline tests/e2e/pageobjects/cell.page.ts:67]: we never re-assign to this variable so we can make it a const
```suggestion
        const commands: NotebookCommand[] = []
```

File: tests/e2e/pageobjects/cell.page.ts [hunk 6]
```diff
@@ -0,0 +1,197 @@
...
+        return text
+    }
+
+    /**
+     * Stalls until the success status check is detected
+     * @param timeout {number} The maximum duration for waiting until it times out
+     * @param expectedContent {string} Override the cell text value in case
+     * the output has some special formatting. (e.g remove new lines from multine content)
+     */
+    async waitForSuccess(expectedContent?: string | undefined, timeout = DEFAULT_SEARCH_FOR_CELL_TIMEOUT) {
```
Comment (christian-bromann) [inline tests/e2e/pageobjects/cell.page.ts:168]: maybe have a more specific name like
```suggestion
    async waitForCellOutput(expectedContent?: string | undefined, timeout = DEFAULT_SEARCH_FOR_CELL_TIMEOUT) {
```

File: tests/e2e/pageobjects/cell.page.ts [hunk 7]
```diff
@@ -0,0 +1,197 @@
...
+        await browser.waitUntil(async () => {
+            return (await this.isSuccessfulExecution(expectedContent)) === true
+        }, {
+            timeout
+        })
+    }
+
+    async areStatusBarCommandsRendered(elements: StatusBarElements[]) {
```
Comment (christian-bromann) [inline tests/e2e/pageobjects/cell.page.ts:176]:
```suggestion
    async isStatusBarCommandsRendered(elements: StatusBarElements[]) {
```

File: tests/e2e/specs/basic.e2e.ts [hunk 8]
```diff
@@ -51,27 +55,230 @@ describe('Runme VS Code Extension', () => {
...
-    await expect($('.output_container').$('>>>pre')).toHaveText('Foo 👀\nBar 🕺\nLoo 🚀')
+    it('more shell example', async () => {
+      const cell = await notebook.getCell('echo "Foo 👀"\nsleep 2\necho "Bar 🕺"\nsleep 2\necho "Loo 🚀"')
+      await cell.run()
+      await cell.waitForSuccess()
+      expect(await cell.cellOutputExists('Foo 👀\nBar 🕺\nLoo 🚀', OutputType.ShellOutput)).toBe(true)
+      expect(await cell.areCommonStatusBarCommandsRendered()).toBe(true)
+    })
+
+
```
Comment (christian-bromann) [inline tests/e2e/specs/basic.e2e.ts:83]:
```suggestion
```

**Review by christian-bromann — COMMENTED** [review]
Some minor comments, overall LGTM

File: examples/README.md [hunk 9]
```diff
@@ -68,7 +68,7 @@ Supports multiple lines where the export is just somewhere in between:
 
 ```sh
 echo "Auth token for service foo"
-export SERVICE_FOO_TOKEN=foobar
```
Comment (christian-bromann) [inline examples/README.md:71]: Setting the env var without quotes was intentional so we can check if Runme can handle both, is this not supported anymore?

File: tests/e2e/pageobjects/cell.page.ts [hunk 10]
```diff
@@ -0,0 +1,229 @@
...
+
+            for await (const row of commandsResult$$) {
+                const text = await row.getText()
+                commands.push({
+                    element$: row,
+                    text: text.trim()
+                })
+            }
+        }
+        return commands
```
Comment (christian-bromann) [inline tests/e2e/pageobjects/cell.page.ts:81]: I prefer to return early and safe indentations, up to you:

```suggestion
        if (!executionRow) {
            return commands
        }

        const statusBar$ = await executionRow.$('.cell-statusbar-container')
        const statusRight$ = await statusBar$.$('.cell-status-right')
        const contributedRight$ = await statusRight$.$('.cell-contributed-items-right')
        const commandsResult$$ = await contributedRight$.$$('.cell-status-item-has-command')

        for await (const row of commandsResult$$) {
            const text = await row.getText()
            commands.push({
                element$: row,
                text: text.trim()
            })
        }

        return commands
```

**Comment by mxsdev** [comment]
Closing in favor of #411

**Pull request closed without merging**